Support CosmosDb dependency tracking

lmolkova opened the pull request:
Cosmos DB SDK ([Microsoft.Azure.Cosmos v3.TODO](https://www.nuget.org/packages/Microsoft.Azure.Cosmos) now supports dependency tracking, even when used in `Direct` (TCP) mode.

You should see `DependencyTelemetry` reported for all public API calls (that involve network communication with the service). If you use `Gateway` (HTTP) mode, you would also see dependencies for underlying HTTP calls. For `Direct` calls, you might see a few nested HTTP dependencies too, but low-level TCP calls are not traced.

Attributes reported by Cosmos SDK can be found here: https://github.com/Azure/azure-cosmos-dotnet-v3/issues/3058. They are populated as custom dimensions.

New dependencies tracing public API calls should have:
- `Target` following `{CosmosDB resource name} | {database name}` pattern, e.g. `cosmos-tracing-demo | my-store`. 
- `Name` follows `{container name} | {operation}` pattern, e.g. `orders | Read`
- `Type` matches Resource Provider namespace (`Microsoft.DocumentDB`)
- And `sdkversion` matches `rdddsaz`.

You can enable distributed tracing for Cosmos with the following API: TODO, add a link to cosmos docs.

In addition to dependency calls, Cosmos DB reports extensive diagnostics when an operation fails or takes too long. You can configure it with the following settings on the Cosmos DB client - TODO.
This diagnostics is sent as a log (`TraceTelemetry`) with `Warning` severity having `CosmosDiagnostics` as JSON in the message. You can identify it using `EventName` (in custom dimensions) with the `RecordDiagnosticsForRequests` value.
Such `TraceTelemetry` should be a child of Cosmos DB dependency telemetry described above.


## Changes

CosmosDB direct (TCP mode) dependency tracking (coming in future versions of Cosmos DB v3 SDK). 
Uses attributes outlined [here](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/3058) and additional cosmos diagnostics is coming from the specific event source and is tracked as `TraceTelemetry`. 

EventSource verbosity is expected to be low - it will be controlled by CosmosDb customers on CosmosDb side. Cosmos would write it when an operation takes longer than a configurable threshold or when an error happens.


### Checklist
- [x] CHANGELOG.md updated with one line description of the fix, and a link to the original issue if available.

### Notes for reviewers:

- We support [comment build triggers](https://docs.microsoft.com/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#comment-triggers)
  - `/AzurePipelines run` will queue all builds
  - `/AzurePipelines run <pipeline-name>` will queue a specific build

lmolkova commented:
/cc @sourabh1007

cijothomas commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticListenerSubscriber.cs:
```diff
@@ -2,14 +2,24 @@
 
         public AzureSdkDiagnosticListenerSubscriber(TelemetryConfiguration configuration) : base(configuration)
         {
+            // listen to Cosmos EventSource only - other logs can be sent using ILogger
```
I understand this is draft PR, but can we just ask customers to leverage existing EventSource adapter to minimize changes here?
 
https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-trace-logs#use-eventsource-events

lmolkova replied:
Good point! I'd love to, but it'd mean users would need to install and configure `Microsoft.ApplicationInsights.EventSourceListener`, while we believe the events coming through this source are essential for users to debug any cosmos-related issues. 
Adding a dependency to `Microsoft.ApplicationInsights.EventSourceListener` from DependencyCollector would be an overkill.

sourabh1007 commented:
Why PR headline talk about Cosmos DB "Direct Mode" only?

lmolkova commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/RemoteDependencyConstants.cs:
```diff
@@ -10,6 +10,7 @@ internal static class RemoteDependencyConstants
         public const string AzureTable = "Azure table";
         public const string AzureQueue = "Azure queue";
         public const string AzureDocumentDb = "Azure DocumentDB";
+        public const string AzureCosmosDb = "Azure CosmosDB";
```
DocumentDb is a legacy name and it'd be great to use CosmosDb instead.

TODO: need to run it by Portal team as well.

lmolkova replied:
Checked with @OsvaldoRosado - we should populate resource provider namespace - `Microsoft.DocumentDB`

sourabh1007 commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticsEventHandler.cs:
```diff
@@ -325,6 +337,49 @@ private static bool IsMessagingDependency(string dependencyType)
+            }
+
+            // similar to SqlClientDiagnosticSourceListener
+            telemetry.Target = string.Join(" | ", dbAccount, dbName);
```
Anyhow Appinsight is removing documents.azure.com. We would need it

![image](https://user-images.githubusercontent.com/6362382/192016630-fae60fd1-3881-4e7e-88c5-1db1734ba361.png)

sourabh1007 replied:
There is value to have this information on UI....Since we have limited space on UI..let me take the team's feedback and come back to you on this.

sourabh1007 commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticListenerSubscriber.cs:
```diff
@@ -2,19 +2,29 @@
         public AzureSdkDiagnosticListenerSubscriber(TelemetryConfiguration configuration) : base(configuration)
         {
+            // listen to Cosmos EventSource only - other logs can be sent using ILogger
+            this.logsListener = new AzureSdkEventListener(this.Client, EventLevel.Warning, "Azure.Cosmos");
```
Can we make Event level user configurable?

lmolkova replied:
yes, we can, it might still be a bit difficult for users to configure it. It seems we also have a configuration on the Cosmos SDK side and it'd be nice to have all control on one SDK side and avoid inconsistent config, what do you think?

sourabh1007 replied:
At SDK side customer cannot control which type of event needs to be listened. 
SDK will emit events on the basis of listener's configuration. So, customer has to configure at AI SDK which kind of events they want to listen, and SDK will emit accordingly.
From SDK customer can only control the feature enablement.

lmolkova replied:
> At SDK side customer cannot control which type of event needs to be listened.
But they control which events are emitted with enablement flag AND thresholds, no? 

Can you please explain the scenario that you want to enable?
customer can't configure Cosmos SDK, but can  configure AI SDK?

AI SDK enablement tends to be one line of code, most customers will struggle to configure things beyond that, especially when they use codeless attach, I'd rather avoid adding extra configuration if there is no good scenario or need.

sourabh1007 replied:
if I am not mistaken, here you are capturing only Event Level as Warning? What will happen if SDK emits events in Event level as Information? will those get captured?

sourabh1007 replied:
another question about this hardcoded "Azure.Cosmos" I am wondering after this PR we won't be able to change this name? I think we should be fine with this as we won't be able to change this name in CosmosDb SDK also.

lmolkova replied:
Discussed with @sourabh1007 offline:
- AI SDK should listen to Info events by default. No config necessary
- Cosmos SDK owns tracing configuration (enablement, threshold to record diagnostics, etc), so it should also allow customers to enable info events.

This way all cosmos tracing configurations will stay in one place and would apply to AI SDK, OTel and custom listeners.

sourabh1007 commented:
Events are loosing message while listened by appinsight..below is the log I got

_Application Insights Telemetry: {"name":"AppTraces","time":"2022-09-23T17:42:26.1395222Z","iKey":"2fabff39-6a32-42da-9e8f-9fcff7d99c6b","tags":{"ai.application.ver":"1.0.0.0","ai.cloud.roleInstance":"DESKTOP-I8MO142","ai.operation.id":"2239023b88bb44b60eab840e1983cded","ai.operation.parentId":"80155111011166ad","ai.internal.sdkVersion":"dotnetc:2.22.0-14","ai.internal.nodeName":"DESKTOP-I8MO142"},"data":{"baseType":"MessageData","baseData":{"ver":2,**"message":"n/a"**,"severityLevel":"Warning","properties":{"EventName":"WriteWarningEvent","EventId":"2","AspNetCoreEnvironment":"Development","DeveloperMode":"true","CategoryName":"Azure.Cosmos"}}}}_

Cosmosdb SDK doesn't send n/a.

lmolkova commented:
> Events are loosing diagnostics message while listened by appinsight..below is the log I got

Fixed, please check

sourabh1007 commented:
with latest changes, this are looking good except one minor thing i.e. I am seeing cosmosdb text instead of cosmosdb icon (like we were getting previously)

![image](https://user-images.githubusercontent.com/6362382/192577641-c95eaf4c-8193-42c7-8987-204594fa8b0b.png)

lmolkova commented:
> with latest changes, things are looking good except one minor thing i.e. I am seeing cosmosdb text instead of cosmosdb icon (like we were getting previously) 

for this we need portal team to support Azure CosmosDB (in addition to DocumentDB)  - there is nothing I can do on SDK side

sourabh1007 commented:
> > with latest changes, things are looking good except one minor thing i.e. I am seeing cosmosdb text instead of cosmosdb icon (like we were getting previously)
> 
> for this we need portal team to support Azure CosmosDB (in addition to DocumentDB) - there is nothing I can do on SDK side

but previously this icon was coming. in this latest package only it is not showing up

sourabh1007 reviewed (approved):
Looks fine to me. Thanks :)

reyang commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkEventListener.cs:
```diff
@@ -0,0 +1,117 @@
+        private readonly EventLevel level;
+        private readonly string traitName;
+
+        public AzureSdkEventListener(TelemetryClient telemetryClient, EventLevel level, string traitName)
```
Minor, maybe s/traitName/prefix?

reyang reviewed (approved):
LGTM.

cijothomas commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticListenerSubscriber.cs:
```diff
@@ -2,19 +2,29 @@
         public AzureSdkDiagnosticListenerSubscriber(TelemetryConfiguration configuration) : base(configuration)
         {
+            // listen to Cosmos EventSource only - other logs can be sent using ILogger
+            this.logsListener = new AzureSdkEventListener(this.Client, EventLevel.Informational, "Azure.Cosmos");
```
Are these logs fired very rarely? or they can be verbose? Since this is going to be enabled by default, we'll need ensure it wont be too noisy.

cijothomas replied:
just saw that PR description answers this qn already..

cijothomas commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticsEventHandler.cs:
```diff
@@ -325,6 +333,49 @@ private static bool IsMessagingDependency(string dependencyType)
+            }
+
+            // similar to SqlClientDiagnosticSourceListener
+            telemetry.Target = string.Join(" | ", dbAccount, dbName);
```
ever need to account for the possibility of either of this being null?

lmolkova replied:
it should not be possible, but agree that we should be a bit more cautions here. fixed.

cijothomas commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticsEventHandler.cs:
```diff
@@ -325,6 +333,49 @@ private static bool IsMessagingDependency(string dependencyType)
+
+            // similar to SqlClientDiagnosticSourceListener
+            telemetry.Target = string.Join(" | ", dbAccount, dbName);
+            telemetry.Name = string.Join(" | ", dbContainer, dbOperation);
```
same as above - do we need to anticipate this being null?

cijothomas commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticsEventHandler.cs:
```diff
@@ -325,6 +333,49 @@ private static bool IsMessagingDependency(string dependencyType)
+                    continue;
+                }
+                
+                telemetry.Properties[tag.Key] = tag.Value;
```
if we are already populating db.name, and other things into target/name fields, do we need to store them again into property bag as well?

lmolkova replied:
Since `Target` and `Name` contain concatenated values optimized for portal UX and specific for AppInsights, we want to keep the original attributes as is for custom queries - e.g. we want users to be able to query by operation across containers or across all databases in account. This way Cosmos can document their telemetry once and it'd work for any backend: Azure Monitor, Jaeger, or Splunk.

cijothomas commented on WEB/Src/DependencyCollector/DependencyCollector/Implementation/AzureSdk/AzureSdkDiagnosticsEventHandler.cs:
```diff
@@ -15,6 +15,7 @@
 
     internal class AzureSdkDiagnosticsEventHandler : DiagnosticsEventHandlerBase
     {
+        private const string CosmosDBResourceProviderNs = "Microsoft.DocumentDB";
```
Is DocumentDB instead of CosmosDB intentionally done for UI to show icons? If yes, please add a small comment above.

lmolkova replied:
Yes, this is intentional, comment added

cijothomas reviewed (approved):
LGTM, left few minor comments.
Given we do not have true e2e tests, would you update the PR description (or a new issue), to act as the documentation for this enhancement? To answer potential customer/support questions like:
What is the min version of CosmosDb package required?
Whether this works for TCP mode as well? Would we see duplicate if Http mode is used?
Also add a note that TraceTelemetry is produced for certain scenarios (this is unusual compared to other dependencies collected automatically)
How are target/name populated?

lmolkova commented:
Thank you @reyang and @cijothomas for the review! I believe I addressed all the feedback in the code. PTAL.

As for documentation, I updated the description on this PR, just need to add a few details.

@sourabh1007, can you please help us clarify the following:
-  Exact Cosmos DB version that supports tracing
- Is there an example on how to enable and configure tracing? Can you share it or provide a link to existing samples/docs, anything? 

Thanks!
